### PR TITLE
[CAS-1491] Fix hard delete message

### PR DESCRIPTION
**Pull request by bychkovdmitry**
https://stream-io.atlassian.net/browse/CAS-1491

### 🎯 Goal

Completely remove hard deleted messages from our persistent storage, cache and the UI.

### 🎨 UI Changes

<table>
<thead>
<tr>
<th>Before</th>
<th>After</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<video src="https://user-images.githubusercontent.com/9600921/145586533-24153bb4-a2e3-42db-9949-f4ad411a638e.mov" controls="controls" muted="muted" />
</td>
<td>
<video src="https://user-images.githubusercontent.com/9600921/145584814-056e948a-8a31-4d0f-b679-4276343efcca.mov" controls="controls" muted="muted" />
</td>
</tr>
</tbody>
</table>

### 🧪 Testing

- Hard delete a message from another client
- Observe that the message is completely wiped out from the channel

### ☑️ Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] PR targets the `develop` branch
- [x] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [x] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS (task created))
- [x] Reviewers added

### CODE REVIEW

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/channel/ChannelController.kt [hunk 1]
```diff
@@ -738,7 +738,11 @@ public class ChannelController internal constructor(
                 channelLogic.setHidden(false)
             }
             is MessageDeletedEvent -> {
-                upsertEventMessage(event.message)
+                if (event.hardDelete) {
+                    removeLocalMessage(event.message)
+                } else {
+                    upsertEventMessage(event.message)
+                }
                 channelLogic.setHidden(false)
```
Comment (bychkovdmitry) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/channel/ChannelController.kt:746]: So, this part is trivial

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/EventHandlerImpl.kt [hunk 2]
```diff
@@ -392,6 +392,12 @@ internal class EventHandlerImpl(
                     domainImpl.repos.deleteChannelMessagesBefore(event.cid, event.createdAt)
                     domainImpl.repos.setChannelDeletedAt(event.cid, event.createdAt)
                 }
+                is MessageDeletedEvent -> {
+                    if (event.hardDelete) {
+                        domainImpl.repos.deleteChannelMessage(event.message)
+                        domainImpl.repos.evictChannelMessage(event.cid, event.message.id)
+                    }
```
Comment (bychkovdmitry) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/EventHandlerImpl.kt:399]: And this guy is less obvious. Basically it resets 3 caches:
1. In memory cache in channel repository.
2. In-memory cache in message repository. 
3. Database cache in message repository.

File: stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/repository/domain/channel/ChannelRepository.kt [hunk 3]
```diff
@@ -109,6 +110,13 @@ internal class ChannelRepositoryImpl(
         }
     }
 
+    override suspend fun evictChannelMessage(cid: String, messageId: String) {
+        val channel = channelCache.get(cid)
+        if (channel != null) {
+            channel.messages = channel.messages.filter { it.id != messageId }
+        }
+    }
+
```
Comment (bychkovdmitry) [inline stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/repository/domain/channel/ChannelRepository.kt:119]: If we don't do this, the following will happen:
1. Message list screen
2. A message is hard deleted
3. The message disappears
4. Channel list screen
5. Message list screen
6. Deleted message is displayed (Messages from message repository and channel repository are concatenated)

Comment (ogkuzmin) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/repository/domain/channel/ChannelRepository.kt:119]: What about just remove such channel from the cache? It make logic simpler and easier. Next time when someone will fetch channel by id the repo takes it from DB

Comment (bychkovdmitry) [reply stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/repository/domain/channel/ChannelRepository.kt:116]: Thanks, great suggestion!

**Review by ogkuzmin — APPROVED** [review]

**Pull request merged**